fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/IndexAggregateFunctionCall.java

```diff
@@ -0,0 +1,269 @@
+/*
+ * IndexAggregateFunctionCall.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.metadata;
+
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.record.metadata.expressions.EmptyKeyExpression;
+import com.apple.foundationdb.record.metadata.expressions.GroupingKeyExpression;
+import com.apple.foundationdb.record.metadata.expressions.KeyExpression;
+import com.apple.foundationdb.record.query.expressions.AndComponent;
+import com.apple.foundationdb.record.query.expressions.BaseField;
+import com.apple.foundationdb.record.query.expressions.Comparisons;
+import com.apple.foundationdb.record.query.expressions.FieldWithComparison;
+import com.apple.foundationdb.record.query.expressions.NestedField;
+import com.apple.foundationdb.record.query.expressions.QueryComponent;
+import com.apple.foundationdb.record.query.plan.temp.EnumeratingIterable;
+import com.apple.foundationdb.record.query.plan.temp.TopologicalSort;
+import com.google.common.base.Objects;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSetMultimap;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Sets;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+/**
+ * Class to represent a construct {@code AGG by Set.of(col1, col2, ...)}. This is the unbound equivalent of
+ * {@link IndexAggregateFunction} which is bound to an index and represents its group-by columns as a list of
+ * {@link KeyExpression} that match the order imposed by bound index rather than a set.
+ */
+@SuppressWarnings("unused")
+@API(API.Status.INTERNAL)
+public class IndexAggregateFunctionCall {
```

Comment:
Would the name `UnboundIndexAggregateFunction` be appropriate?

Reply:
I guess you can do that. I kind of like function call better due to in my mind being the representation of a call-site of a function. Very much like here -> https://en.wikipedia.org/wiki/Call_site